File: leetcode/easy/145-postorder/stack.cpp
/**
 * @file stack.cpp
 * @author Vortexx2
 * @brief Problem 145 - Binary Tree Postorder Traversal
 * @date 23-07-2021
 * 
 * Runtime - 0 ms
 * Memory Usage - 8.3 MB
 */
#include <iostream>
#include <vector>
#include <deque>
#include <unordered_map>
#include <map>
#include <set>
#include <stack>
#include <queue>
#include <string>

using namespace std;

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  vector<int> postorderTraversal(TreeNode *root)
  {
    vector<int> res;

    if (!root)
      return res;

    stack<TreeNode *> st;

    while (true)
    {
      // Go to the left most point of the current node (root)
      while (root)
      {
        st.push(root);
        root = root->left;
      }

      if (st.empty())
        break;

      root = st.top();
      TreeNode *prev = NULL;
      while ((!root->right) || (root->right == prev))
      {
        res.push_back(root->val);
        st.pop();
        // After pop, accessing st.top() might cause memory issue
        if (st.empty())
          return res;
        prev = root;
        root = st.top();
      }

      root = root->right;
    }

    return res;
  }
};

File: leetcode/easy/145-postorder/stack.py
"""
  @author Vortexx2
  Problem 145 - Binary Tree Postorder Traversal

  Runtime - 32 ms
  Memory Usage - 14.1 MB
"""
from typing import List


# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right


class Solution:
  def postorderTraversal(self, root: TreeNode) -> List[int]:
    if not root:
      return []

    st = []
    res = []

    while True:

      while root:
        st.append(root)
        root = root.left

      if not st:
        return res

      root = st[-1]
      prev = None
      while ((not root.right) or (root.right == prev)):
        res.append(root.val)
        st.pop()

        if not st:
          return res

        prev = root
        root = st[-1]

      root = root.right
